SwiftCompilerSources/Sources/Optimizer/Utilities/AccessUtils.swift

```diff
@@ -0,0 +1,421 @@
+      // case it wouldn't be local
+      return (baseAddress as! ProjectBoxInst).operand is AllocBoxInst
+    case .stack:
+      return true
+    default:
+      return false
+    }
+  }
+
+  /// Returns `true` if we can reliably compare for equality this `AccessBase`
+  /// with another `AccessBase`.
+  /// When comparing two uniquely identified access bases, it can be followed that if they are not equal,
+  /// also the accessed memory addresses do not alias.
+  /// This is e.g. not the case for class references: two different references may still point to the same object.
+  var isUniquelyIdentified: Bool {
+    switch baseKind {
+    case .box:
+      // The operand `%op` in `%baseAddress = project_box %op` can
+      // be `alloc_box` or an argument. Only if it's a fresh allocation it is
+      // uniquelyIdentified, otherwise it is aliasable, as all the other references.
+      return baseAddress is AllocBoxInst
+    case .stack, .global:
+      return true
+    case .argument:
+      // An argument address that is non-aliasable
+      return (baseAddress as! FunctionArgument).isExclusiveIndirectParameter
+    case .class, .tail, .yield, .pointer:
+      // References (.class and .tail) may alias, and so do pointers and
+      // yield results
+      return false
+    }
+  }
+
+  /// Returns `true` if the two access bases do not alias
+  func isDistinct(from other: AccessBase) -> Bool {
+    switch (baseAddress, other.baseAddress) {
+    case is (AllocStackInst, AllocStackInst),
+         is (AllocBoxInst, AllocBoxInst):
+      return baseAddress != other.baseAddress
+    case let (this as GlobalAddrInst, that as GlobalAddrInst):
+      return this.global != that.global
+    case let (this as FunctionArgument, that as FunctionArgument):
+      return (this.isExclusiveIndirectParameter || that.isExclusiveIndirectParameter) && this != that
+    case let (this as RefElementAddrInst, that as RefElementAddrInst):
+      return (this.fieldIndex != that.fieldIndex)
+    default:
+      if isUniquelyIdentified && other.isUniquelyIdentified && baseKind != other.baseKind { return true }
+      // property: `isUniquelyIdentified` XOR `isObject`
+      if isUniquelyIdentified && other.baseKind.isObject { return true }
+      if baseKind.isObject && other.isUniquelyIdentified { return true }
+      return false
+    }
+  }
+}
+
+/// An `AccessPath` is a pair of a `base: AccessBase` and a `projectionPath: Path`
+/// which denotes the offset of the access from the base in terms of projections.
+struct AccessPath : CustomStringConvertible {
+  var base: AccessBase
+
+  /// address projections only
+  var projectionPath: SmallProjectionPath
+
+  var description: String {
+    "\(projectionPath): \(base)"
+  }
+
+  func isDistinct(from other: AccessPath) -> Bool {
+    return base.isDistinct(from: other.base) ||
+      (base.baseAddress == other.base.baseAddress && projectionPath != other.projectionPath)
```

Comment:
Comparing the paths is wrong, because one path could e.g. be a sub-path of the other, i.e. paths are not equal, but overlapping. Or the paths could contain wildcards.
I suggest that either you do a correct "mayOverlap" check for projection paths or you remove this function at all.